core/src/io/anuke/mindustry/graphics/BlockRenderer.java

```diff
@@ -269,15 +273,53 @@ void cacheChunk(int cx, int cy, boolean floor){
 		cbatch.end();
 		cache[cx][cy][floor ? 0 : 1] = cbatch.getLastCache();
 	}
-
+	
 	public void clearTiles(){
 		cache = null;
 		createBatch();
 	}
-
+	
 	private void createBatch(){
 		if(cbatch != null)
 			cbatch.dispose();
 		cbatch = new CacheBatch(world.width() * world.height() * 4);
 	}
-}
+	
+	public void drawPreview(Block block, float drawx, float drawy, float rotation, float opacity) {
+		Draw.reset();
+		Draw.alpha(opacity);
+		Draw.rect(block.name(), drawx, drawy, rotation);
+	}
+	
+	public void handlePreview(Block block, float rotation, float drawx, float drawy, int tilex, int tiley) {
+		
+		if(control.input().recipe != null && state.inventory.hasItems(control.input().recipe.requirements)
+		   && control.input().validPlace(tilex, tiley, block) && (android || control.input().cursorNear())) {
+			
+			float opacity = (float)Settings.getInt("previewopacity")/100f;
+			
+			if(block.isMultiblock()) {
+				if((tiley - control.input().getBlockY()) % block.height != 0
+				   || (tilex - control.input().getBlockX()) % block.width != 0) return;
+			}
+			
+			if(block instanceof Turret) {
+				Draw.alpha(opacity);
+				if (block.isMultiblock()) {
+					Draw.rect("block-" + block.width + "x" + block.height, drawx, drawy);
+				} else {
+					Draw.rect("block", drawx, drawy);
+				}
+			}
+			
+			drawPreview(block, drawx, drawy, rotation, opacity);
+			
+			Tile tile = world.tile(tilex, tiley);
+			if((block instanceof Drill || block instanceof Pump) && block.isLayer(tile)) {
```

Comment:
This is rather hacky, and will break with multiblock drills, pumps, or other new blocks that display a red slash. I don't think it's necessary.